plugins/module_utils/copy.py

```diff
@@ -0,0 +1,339 @@
+        ``name`` (string), ``size`` (integer), ``mode`` (integer, see https://pkg.go.dev/io/fs#FileMode),
+        ``mtime`` (string), and ``linkTarget`` (string);
+    :link_target: is ``None`` if the file is not a symlink or when ``follow_links=False``,
+        and a string with the symlink target otherwise.
+    """
+    considered_in_paths = set()
+
+    while True:
+        if in_path in considered_in_paths:
+            raise DockerFileCopyError('Found infinite symbolic link loop when trying to stating "{0}"'.format(in_path))
+        considered_in_paths.add(in_path)
+
+        if log:
+            log('FETCH: Stating "%s"' % in_path)
+
+        def f(client):
+            response = client._head(
+                client._url('/containers/{0}/archive', container),
+                params={'path': in_path},
+            )
+            if response.status_code == 404:
+                return None
+            client._raise_for_status(response)
+            header = response.headers.get('x-docker-container-path-stat')
+            try:
+                return json.loads(base64.b64decode(header))
+            except Exception as exc:
+                raise DockerUnexpectedError(
+                    'When retrieving information for {in_path} from {container}, obtained header {header!r} that cannot be loaded as JSON: {exc}'
+                    .format(in_path=in_path, container=container, header=header, exc=exc)
+                )
+
+        stat_data = call_client(f)
+        if stat_data is None:
+            return in_path, None, None
+
+        # https://pkg.go.dev/io/fs#FileMode: bit 32 - 5 means ModeSymlink
+        if stat_data['mode'] & (1 << (32 - 5)) != 0:
+            link_target = stat_data['linkTarget']
+            if not follow_links:
+                return in_path, stat_data, link_target
+            in_path = os.path.join(os.path.split(in_path)[0], link_target)
+            continue
+
+        return in_path, stat_data, None
+
+
+def fetch_file_ex(call_client, container, in_path, process_none, process_regular, process_symlink, follow_links=False, log=None):
+    """Fetch a file (as a tar file entry) from a Docker container to local."""
+    considered_in_paths = set()
+
+    while True:
+        if in_path in considered_in_paths:
+            raise DockerFileCopyError('Found infinite symbolic link loop when trying to fetch "{0}"'.format(in_path))
+        considered_in_paths.add(in_path)
+
+        if log:
+            log('FETCH: Fetching "%s"' % in_path)
+        try:
+            stream = call_client(
+                lambda client: client.get_raw_stream(
+                    '/containers/{0}/archive', container,
+                    params={'path': in_path},
+                    headers={'Accept-Encoding': 'identity'},
+                )
+            )
+        except DockerFileNotFound:
+            return process_none(in_path)
+
+        # TODO: stream tar file instead of downloading it into a BytesIO
```

Comment:
I see the note in the doc about not using this for large files; it would be nice to have it use streams so arbitrarily large files could be supported.

Reply:
I've been playing around with streaming a bit, I think the easiest way is to not use Python's `tarfile`, but instead use some own code that allows to stream the contents of the generated tarfile. I'll play around with it some more...

Reply:
I think this should be doable without too much repetition; `tarfile.TarInfo` does part of the heavy lifting (i.e. composing the file header) and can be used separately.

Reply:
Ok, I've implemented that in 052c7f6214a619768ffe289fe106246918cc5eff.